### PR TITLE
Move SeaBios debug log to be seen as part of the virt-launcher log

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
As @vladikr suggested in a [previous PR](https://github.com/kubevirt/kubevirt/pull/4917#discussion_r576464454), SeaBios debug logs should be part of virt-launcher log.

Also, as [explained here](https://www.seabios.org/Debugging#Timing_debug_messages) use a pipe instead of a file for logging to prevent the creation of a log file that keeps getting larger.

**Special notes for your reviewer**:
For some reason "syscall.Mkfifo" doesn't make a pipe with the permissions passed as a parameter. I've found a [solution here](https://wenzr.wordpress.com/2018/03/27/go-file-permissions-on-unix/) which is basically to make another call to "syscall.Chmod" to fix pipe's permissions.

Am I missing something here? Anyways I would be happy to hear about a more elegant solution.

**Release note**:

```release-note
QEMU SeaBios debug logs are being seen as part of virt-launcher log.
```

### CODE REVIEW

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: I'm not sure I understand the logic here. What happen if Stat() returns some error which is not IsNotExist  ? It seems like you just continue as usual (or I missed something)

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: Do we really need two separate consts?  The other is used only for printing and even there maybe it's better to print the full path ?

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: Any special reason for panic() and not return of error ?

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (iholder-redhat) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: Sorry, it's a left-over. I'll remove it. Thanks!

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (iholder-redhat) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go]: Sure, I'll print the full path. thanks

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:382]: I don't think you should explicitly close the pipe. I think it's being closed by the `wait()`

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:394]: What is the usage of this sleep ?

File: cmd/virt-launcher/readserial.py
Comment (vladikr) [inline cmd/virt-launcher/readserial.py:1]: I'm missing something here... 
why do we need to introduce python into KubeVirt?
Can we write whatever this script is doing in go?
It's not difficult to `select` in go...

**Comment by vladikr** [comment]
@iholder-redhat why do we need to have this python script? We can easily handle this with go.

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (iholder-redhat) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:382]: Are you sure? I'm not sure that's right. I wait for "readSerialScriptCmd" which only consumes from the pipe by executing a python script. The pipe itself was created though Mkfifo above.

**Comment by iholder-redhat** [comment]
@vladikr Is that what you want? Sure we can handle this with Go, it's just that you provided [this link](https://www.seabios.org/Debugging#Timing_debug_messages) and wrote "we could actually do it better" so it seemed like you meant using their python script.

I'll re-write the script logic into a Go function and use it instead.

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:382]: Maybe I'm off here, but I thought it's:
your script  indeed consumes from the QEMUSeaBiosDebugPipe  pipe , and then you create another pipe cmdStdout for the script output (using StdoutPipe() ).   The cmdStdout  should be closed by the wait

Comment (ezrasilvera) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:382]: Maybe you wanted to close QEMUSeaBiosDebugPipe  ?

**Comment by ezrasilvera** [comment]
> @vladikr Is that what you want? Sure we can handle this with Go, it's just that you provided [this link](https://www.seabios.org/Debugging#Timing_debug_messages) which explains how to use their python script for timing debug messages and wrote "we could actually do it better" so it seemed like you meant using their python script.
> 
> I'll re-write the script logic into a Go function and use it instead.

I think if it's not too complicated it's preferable.  You can also just take the needed functionality from that script ..

**Comment by iholder-redhat** [comment]
@ezrasilvera Sure, I'll do that and re-push.

/hold

**Comment by vladikr** [comment]
> @vladikr Is that what you want? Sure we can handle this with Go, it's just that you provided [this link](https://www.seabios.org/Debugging#Timing_debug_messages) which explains how to use their python script for timing debug messages and wrote "we could actually do it better" so it seemed like you meant using their python script.
> 
> I'll re-write the script logic into a Go function and use it instead.

@iholder-redhat sorry if I confused you. With this link, I simply wanted to show an example of how we could handle this.
I'd prefer if we didn't introduce any additional scripts into the project if we could do the same natively and test it.

**Comment by iholder-redhat** [comment]
@vladikr Tha's alright :) sorry for the misunderstanding. I've removed the script. The Go implementation turned out to be much simpler.

/hold cancel

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:365]: Can you please explain what exactly the timing info you give here and why is it important?  
Why do we need delta from some starting point, why is that starting point being reset if the diff is > 60 sec? 
And in general why not just putting a timestamp?

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:353]: nit: why not using the built in time functions like `time.Since(start)`

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (iholder-redhat) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:365]: Basically I've tried to mimic the script's logic. I did that in the first place because Vladik said to do so by [following this explanation](https://www.seabios.org/Debugging#Timing_debug_messages).

@vladikr Can you answer that question?

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:259]: I think you can now remove the "nosec" comment line. This is used to skip gosec validation , you no longer have now. (you now use constant strings and not variables in exec)

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:284]: Why not printing the actual case i.e., "failed to open  <logfile>"

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:338]: Maybe move this to after the error handling?  Otherwise defer close() might try to close a not opened filed

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:284]: wouldn't `Reason(err)` log the same error as err.Error()? are we logging it twice?

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: I got confused here :)
do you read this twice on purpose ? `line, _, err = reader.ReadLine()`

**Comment by vladikr** [comment]
Overall, everything looks good. I only had a couple of questions.

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (iholder-redhat) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:284]: Makes sense. I wasn't aware that that's what `Reason(err)` is doing. Thanks :)

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (iholder-redhat) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: This wasn't my intention. I thought what happens is (similar to C++/Java for loops):
* Before first iteration: `line, _, err` are being initialized with `reader.ReadLine()`
* After each iteration break if `err != nil`
* Before starting a new iteration, update `line, _, err` by calling `reader.ReadLine()`.

So as I see it `line` isn't being read twice, but once for initialization and another time for each iteration.
Or am I missing something here?

Comment (ezrasilvera) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: @iholder-redhat I don't think your for loop is doing what you intended. In general the golang loop is like this :
```
for [ Initial Statement ] ; [ Condition ] ; [ Post Statement ] {
}
```
And this is not aligned with your assumptions.  I suggest to just do the Redline inside the loop and then check the error, if EOF then break , if another error return, if not an error do the process. 

BTW, are you familiar with golang playground?   If you are not sure about behavior you can easily write a very small piece of code and test it there directly ... For example you could tested the loop behavior

Comment (ezrasilvera) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: @itamarh I think your loop will work correctly (your description is correct), but it's indeed a little confusing :-) 
Usually people will just do the readline inside the loop body, and add explicit tests for error (return) and EOF (break), which might be clearer.

Comment (vladikr) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: I'm just curious why did you choose to use a ReadLine? Wouldn't it be simpler to use a scanner? 
Maybe something like this will work?
```
var line string
 scanner := bufio.NewScanner(file)                                           
    if scanner.Scan() {                                                         
        line = scanner.Text()                                                 
    }                                                                           
    if err := scanner.Err(); err != nil {                                       
        return nil, err                                                         
    }
```

Comment (ezrasilvera) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: @vladikr Don't you need to have `for scanner.Scan()  { `  ?

Comment (vladikr) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: oh yes, of course :)
sorry

Comment (iholder-redhat) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: @ezrasilvera I see what you mean. It is a bit confusing :) It's just that I usually don't prefer loops which lack a concrete condition as it's more error prone as it can lead to infinite loops, i.e.
```
for {
    line, err := ...
    if err != nil {
        break
    }
}
```
Maybe the line read can occur inside the loop (and one time outside of it for initialization) and we'll check the error as a condition for the loop? something like:
```
line, err := ...
for err == nil {
    line, err := ...
}

if err != nil {
    return ...
}
```
IMHO although possible, I'm not sure it's better since, at least in my eyes, it's less obvious to figure out what the for loop is actually doing in the above approach (since the loop body is only error checking).
As I see it what I did is pretty standard - moved the initialization line (`line, err := ...`), the condition (`err == nil`) and the post statement (`line, err := ...`) into the loop body.

Just my opinion :) what do you guys think?

Comment (iholder-redhat) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: @vladikr I can change the implementation to use Scanner. I've read a bit about the differences between Scanner and Reader but found only minor differences while the internet says that for the majority of cases they're interchangeable. As I'm still new to Go can you explain why in your opinion Scanner is better for this case? :) thanks!

Comment (vladikr) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: it's just ReadLine is more low level and as you saw requires more handling. `Scanner` is more general-purpose and incorporates some of this logic.
you can see that the golang page also doesn't recommend ReadLine: https://golang.org/pkg/bufio/#Reader.ReadLine

In any case, I don't have a strong opinion, the look just caught my eye and it took me some time to understand. I'm sure people will start questioning this next time they'll read it :)
I don't want us to stall on this. If it works, either way, I'm +1 :)

Comment (iholder-redhat) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:347]: @vladikr 
I don't have a strong opinion on this either, just wanted to educate myself on the reason for using Scanner instead. If even golang page recommends Scanner I guess it's the right choice. Also I obviously agree that code shouldn't be confusing.

I'll change this to using Scanner and re-upload :) thanks for the answer.

File: pkg/virt-launcher/virtwrap/util/libvirt_helper.go
Comment (ezrasilvera) [inline pkg/virt-launcher/virtwrap/util/libvirt_helper.go:338]: Don't you want to return here ?

Comment (iholder-redhat) [reply pkg/virt-launcher/virtwrap/util/libvirt_helper.go:338]: Yes I do! Thanks for noticing :)

**Comment by ezrasilvera** [comment]
@iholder-redhat Thanks for the PR.
/lgtm

**Comment by vladikr** [comment]
Great!
Thanks!
/approve

**Comment by iholder-redhat** [comment]
> New changes are detected. LGTM label has been removed.

@vladikr This is only a rebase (minor conflict with imports). Could me LGTM me again?
Thanks :)

**Comment by qinqon** [comment]
"""
/hold
until emergency CI fixes of #5207 are merged
"""

**Comment by iholder-redhat** [comment]
> """
> /hold
> until emergency CI fixes of #5207 are merged
> """

I see it is merged :)
/hold cancel
/test pull-kubevirt-e2e-kind-1.17-sriov

**Comment by iholder-redhat** [comment]
/hold
After consulting with @fgimenez we found out that the image for `pull-kubevirt-e2e-kind-1.17-sriov` lane is not updated.

**Comment by fgimenez** [comment]
> After consulting with @fgimenez we found out that the image for `pull-kubevirt-e2e-kind-1.17-sriov` lane is not updated.

in fact the problem is more about the libvirt-go dependency introduced by the changes in this PR, we don't want to spread the need to have specific packages installed in the target platform to be able to build KubeVirt.

**Comment by iholder-redhat** [comment]
> > After consulting with @fgimenez we found out that the image for `pull-kubevirt-e2e-kind-1.17-sriov` lane is not updated.
> 
> thx, in fact the problem is more about the libvirt-go dependency introduced by the changes in this PR, we don't want to spread the need to have specific packages installed in the target platform to be able to build KubeVirt.

@fgimenez The dependency is now removed. Thanks for all the help!!

@ezrasilvera @vladikr What I did is switch the dependency - instead of defining `const QEMUSeaBiosDebugPipe` variable in `pkg/virt-launcher/virtwrap/util/libvirt_helper.go` and import it from `pkg/virt-launcher/virtwrap/converter/converter.go` it is now the opposite: defined in `converter.go` and used in `libvirt_helper.go`. Do you think it's okay? If not - where do you suggest to define `QEMUSeaBiosDebugPipe` variable?

thanks!

**Comment by iholder-redhat** [comment]
> > > After consulting with @fgimenez we found out that the image for `pull-kubevirt-e2e-kind-1.17-sriov` lane is not updated.
> > 
> > 
> > thx, in fact the problem is more about the libvirt-go dependency introduced by the changes in this PR, we don't want to spread the need to have specific packages installed in the target platform to be able to build KubeVirt.
> 
> @fgimenez The dependency is now removed. Thanks for all the help!!
> 
> @ezrasilvera @vladikr What I did is switch the dependency - instead of defining `const QEMUSeaBiosDebugPipe` variable in `pkg/virt-launcher/virtwrap/util/libvirt_helper.go` and import it from `pkg/virt-launcher/virtwrap/converter/converter.go` it is now the opposite: defined in `converter.go` and used in `libvirt_helper.go`. Do you think it's okay? If not - where do you suggest to define `QEMUSeaBiosDebugPipe` variable?
> 
> thanks!

/hold cancel

**Review by dhiller — APPROVED** [review]

**Pull request merged**
